### PR TITLE
Make coverage, like, really fast

### LINKED ISSUE #23157: Running unit tests with coverage enabled takes up to 10x times longer than without coverage
I have a test suite of 150 tests (unit tests, and widget tests). If I run them **without coverage** enabled (`flutter test`), they take less than **10 seconds** to run; however, if I **enabled coverage**, (`flutter test --coverage --coverage-path ./coverage/lcov.info`), then it takes around **1 minute and 40 seconds**.

Flutter doctor:

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel beta, v0.8.2, on Mac OS X 10.13.6 17G65, locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
[✓] iOS toolchain - develop for iOS devices (Xcode 9.2)
[✓] Android Studio (version 3.2)
[✓] IntelliJ IDEA Community Edition (version 2018.1.1)
[✓] VS Code (version 1.28.1)
[✓] Connected devices (1 available)

• No issues found!
```
I am attaching the `--verbose` reports for more information:

[test_report_no_coverage.txt](https://github.com/flutter/flutter/files/2483477/test_report_no_coverage.txt)
[test_report_with_coverage_enabled.txt](https://github.com/flutter/flutter/files/2483478/test_report_with_coverage_enabled.txt)

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Applies changes described in go/better-dart-coverage

## Related Issues

Fixes https://github.com/flutter/flutter/issues/23157

## Tests

I added the following tests:

*Replace this with a list of the tests that you added as part of this PR. A change in behaviour with no test covering it
will likely get reverted accidentally sooner or later. PRs must include tests for all changed/updated/fixed behaviors. See [Test Coverage].*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [ ] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_tools/lib/src/test/coverage_collector.dart [hunk 1]
```diff
@@ -141,3 +154,120 @@ class CoverageCollector extends TestWatcher {
     return true;
   }
 }
+
+Future<Map<String, dynamic>> collect(Uri serviceUri, bool Function(String) libraryPredicate) async {
+  VMService vmService;
+  try {
+    vmService = await VMService.connect(serviceUri, compression: CompressionOptions.compressionOff);
+    await vmService.getVM();
+  } on TimeoutException {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/test/coverage_collector.dart]: This try-catch seems like a no-op.

File: packages/flutter_tools/lib/src/test/coverage_collector.dart [hunk 2]
```diff
@@ -141,3 +154,120 @@ class CoverageCollector extends TestWatcher {
...
+      }
+    }
+    hitMaps.forEach((Uri uri, Map<int, int> hitMap) {
+      coverage.add(_toScriptCoverageJson(uri, hitMap));
+    });
+  }
+  return <String, dynamic>{'type': 'CodeCoverage', 'coverage': coverage};
+}
+
+List<int> _lineAndColumn(int position, List<dynamic> tokenPositions) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/test/coverage_collector.dart]: This method could use some comments.

File: packages/flutter_tools/lib/src/test/coverage_collector.dart [hunk 3]
```diff
@@ -141,3 +154,120 @@ class CoverageCollector extends TestWatcher {
...
+      );
+      futures.add(
+        isolateRef.invokeRpcRaw('getObject', params: <String, dynamic>{'isolateId': isolateRef.id, 'objectId': scriptId})
+          .then((Map<String, dynamic> script) {
+            scripts[scriptId] = script;
+          })
+      );
+    }
+    await Future.wait(futures);
+    final Map<Uri, Map<int, int>> hitMaps = <Uri, Map<int, int>>{};
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/test/coverage_collector.dart]: To help with readability you might split this code up into some more functions.

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/test/coverage_collector.dart]: I refactored this into a method for loading and a method for building

File: packages/flutter_tools/lib/src/test/coverage_collector.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/test/coverage_collector.dart]: Added a description and a link to the vmservice documentation which describes the format

File: packages/flutter_tools/lib/src/test/coverage_collector.dart [hunk 4]
```diff
@@ -141,3 +153,136 @@ class CoverageCollector extends TestWatcher {
...
+  Map<String, Map<String, dynamic>> sourceReports,
+  List<Map<String, dynamic>> coverage,
+) {
+  final Map<Uri, Map<int, int>> hitMaps = <Uri, Map<int, int>>{};
+  for (String scriptId in scripts.keys) {
+    final Map<String, dynamic> sourceReport = sourceReports[scriptId];
+    for (Map<String, dynamic> range in sourceReport['ranges']) {
+      final Map<String, dynamic> coverage = range['coverage'];
+      final Map<String, dynamic> scriptRef = sourceReport['scripts'][range['scriptIndex']];
+      final Uri uri = Uri.parse(scriptRef['uri']);
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/test/coverage_collector.dart]: Do you have to convert scriptRef['uri'] to a Uri object? It looks like all that's done with it is converting it back to a string in _toScriptCoverageJson.

**Review by zanderso — APPROVED** [review]
lgtm w/ question.

**Comment by jonahwilliams** [comment]
Build is currently green - the dashboard may be unable to update older PRs

**Pull request merged**